Implement array.prototype.flat and flatMap

rhuanjl opened the pull request:
This PR implements the stage 3 proposal Array methods: Array.prototype.flat and Array.prototype.flatMap as JsBuiltIns.

Draft specification text is here: https://tc39.github.io/proposal-flatMap/

Notes:
1. 3 test262 failures:
     - length is not configurable (for both methods), this is due to ScriptFunction lengths in CC all not being configurable and should be fixed if #5405 is merged
    - the case using null as this fails -- EDIT: FIXED
2. I can't find a feasible way to test the length is too big error - not sure if CC can actually handle arrays that big anyway so maybe not worth implementing that case?
3. Should probably have a few more CI tests

Fixes #5543

MSLaguana commented on lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js:
```diff
@@ -238,5 +241,157 @@
+            // b. Let exists be ? HasProperty(source, P).
+            if (sourceIndex in source) {
+                // c. If exists is true, then
+                //  i. Let element be ? Get(source, P).
```
Does the code as written work correctly in the presence of proxies / getters? The spec seems to say that the element is accessed once, while below you access it multiple times.

rhuanjl replied:
Should now be fixed - I've also added a case using a Proxy to test this.

ljharb replied:
It’d be great to contribute those tests back to test262 :-)

rhuanjl replied:
@ljharb I'll endeavour to contribute some equivalent tests to test262 later this week - I've not written a test262 test before so will take me a bit of time to work out how to.

MSLaguana commented on lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js:
```diff
@@ -238,5 +241,157 @@
+                //      1. Let elementLen be ? ToLength(? Get(element, "length")).
+                //      2. Set targetIndex to ? FlattenIntoArray(target, element, elementLen, targetIndex, depth - 1).
+                if (__chakraLibrary.isArray(element)) {
+                    targetIndex = __chakraLibrary.FlattenIntoArray(target, element, element.length |0, targetIndex, 0);
```
Huh, just read the spec. I didn't expect that from my intuition about these functions, but I agree with you. While adding the comment, I'd also appreciate one stating that the `thisArg` of the spec is expected to be bound to the `mapperFunction` by the caller.

rhuanjl replied:
Also per spec the mapper function is not passed downwards with the second call - Array.flatMap(func) is functionally equivalent(for simple cases, no getters/proxies etc.) to: Array.map(func).flat().

MSLaguana commented:
Regarding testing the length-too-long error; I'd be surprised if any engine today can hit that case... `2^53-1` is a lot of items! If each just took up 1 bit, you'd still need ~`2^50` bytes of memory... I guess if you were dedicated enough and willing to use swap space then you could get there... But for now I think we're ok without a test for that specific case.

MSLaguana commented:
Regarding the failure with passing `null` as `thisArg`, looks like that might be due to a spec change; see the note at https://tc39.github.io/ecma262/#sec-function.prototype.call

@zenparsing do you know if binding `null` or `undefined` is supposed to keep that value as-is in the called function, rather than substituting the global object?

MSLaguana commented:
Oh; null and undefined aren't replaced if the function is a strict function. I see now that the test case says that it should only be run in strict mode, how were you running it @rhuanjl ? Were you using the test 262 test harness, or just running that one file?

rhuanjl commented:
I made my own mini-harness to run the test files. That case still failed with strict mode though - without strict mode most of the cases in the file failed, with strict mode only that case (the one with null) failed.

boingoing commented on lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js:
```diff
@@ -238,5 +241,157 @@
+                    //  1. Let elementLen be ? ToLength(? Get(element, "length")).
+                    //  2. Set targetIndex to ? FlattenIntoArray(target, element, elementLen, targetIndex, depth - 1).
+                    element = source[sourceIndex];
+                    targetIndex = __chakraLibrary.FlattenIntoArray(target, element, element.length |0, targetIndex, depth - 1);
```
I believe this operation (`element.length |0`) is not doing the same thing as ToLength. Seems the operation is closer to ToInteger since it does not clamp negative values to 0 like ToLength does.

boingoing commented on lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js:
```diff
@@ -238,5 +241,157 @@
+                //  iii. Let shouldFlatten be false.
+                //  iv. If depth > 0, then
+                //      1. Set shouldFlatten to ? IsArray(element).
+                if (depth > 0 && __chakraLibrary.isArray(source[sourceIndex])) {
```
As mentioned above, we should not be doing this property get for `source[sourceIndex]` here as well as another lookup in the conditional branches.

MSLaguana commented on lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js:
```diff
@@ -238,5 +241,157 @@
+        //5. Let A be ? ArraySpeciesCreate(O, 0).
+        const A = __chakraLibrary.arraySpeciesCreate(o, 0);
+        //6. Perform ? FlattenIntoArray(A, O, sourceLen, 0, depthNum).
+        if (!thisArg) {
```
I think this is the reason the test is failing; you are just checking the truthiness. Undefined works because when you call a strict function `foo()` it will get `undefined` as its this, but with this check calling passing false or 0 or the empty string or any other falsy value would fail I'd suggest changing this to `thisArg === undefined`, and in all other cases going with the bind in the else.

rhuanjl replied:
Updated

boingoing commented on lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js:
```diff
@@ -238,5 +241,157 @@
+                    targetIndex = __chakraLibrary.FlattenIntoArray(target, element, element.length |0, targetIndex, 0);
+                } else {
+                    //  vi. Else,
+                    //      1. If targetIndex >= 253-1, throw a TypeError exception.
```
Little typo here, should probably be 2^53-1 not 253-1.

MSLaguana commented:
Apart from those small comments, this looks reasonable to me. Thanks for taking this on!

I'm not sure why the no_icu_static_ubuntu_linux_debug test is failing, it looks like somehow the two functions haven't been added to the array prototype, but in the normal static_ubuntu_linux_debug CI run the test did run and pass.

MSLaguana commented:
Ah, and just saw that windows 10 ci_lite_x64_debug also failed in the same way.

rhuanjl commented:
Do the windows lite build and ubuntu no_icu builds build without JsBuiltins? (these functions are the first JsBuiltins without a fall back C++ implementation so if JsBuiltins are disabled these won't work)

rhuanjl commented:
Thank you for the reviews @MSLaguana @boingoing I think I've fixed all the points you've raised - I'll hold off on squashing for now as I anticipate that some further reviews/changes will be required.

I'm nervous about non-icu/lite builds not supporting this - could be confusing for some users, should/could we either:
a) enable JsBuiltIns for non-icu/lite builds - they don't need icu etc. OR
b) put in a stub that throws a message saying that these methods aren't supported in a non-icu/lite build. OR
c) Make a c++ implentation

rhuanjl commented:
Lite build seems to work with the "normal" no-jit bytecode - tested offline and CI runs with it - as can be seen. I figure that a change of that kind should perhaps be a separate PR though? (I added it here just to test the idea)

sigatrev commented on lib/Common/CommonDefines.h:
```diff
@@ -132,9 +132,10 @@
-#define ENABLE_JS_BUILTINS                          // Built In functions support
 #endif
 
+#define ENABLE_JS_BUILTINS                          // Built In functions support
```
I assume it's historical, so that we had a way to back out if needed. Js Builtins are relatively new with the first one, indexOf, it was not immediately clear if the benefits were significant.

sigatrev commented:
Thanks for taking the time to implement this.

I took a lot at the JIT output and it seems to be doing a reasonable job, but recussions isn't handled very precisely when it comes to inlining. It'd be interesting to see how perf would compare to an implementation using a stack instead of recursion, but that's not strictly necessary for this PR.

rhuanjl commented:
@sigatrev Thanks for the review. I've:

1. removed the IsCallable function AND
2. as a separate commit I've refactored FlattenIntoArray to use a stack - kept as a separate commit so can be easily dropped if not wanted.
3. I've added an extra loop to FlattenIntoArrayMapped so it no longer has to call FlattenIntoArray

sigatrev commented:
@rhuanjl I did some microbenchmarking and the original recursive case runs ~30% faster. Thanks for giving that a shot, and the quick turn-around.

rhuanjl commented:
OK, I've updated to remove the stack based mechanism, I've kept the inner loop for FlattenIntoArrayMapped though as I assume that has minimal performance impact and if anything makes the code clearer.

sethbrenith commented on lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js:
```diff
@@ -238,5 +241,178 @@
+            sourceLen =  __chakraLibrary.GetLength(o);
+        }
+        //3. If IsCallable(mapperFunction) is false throw a TypeError exception
+        if (typeof mapperFunction != "function") {
```
Nit: `!==` is generally preferable

rhuanjl replied:
I copied that from Array.prototype.filter above - so if changing this one will change filter as well for consistency.

sethbrenith replied:
Seems reasonable to me, thanks!

---
In reply to: [208360161](https://github.com/Microsoft/ChakraCore/pull/5573#discussion_r208360161) [](ancestors = 208360161)

dilijev commented on lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js:
```diff
@@ -238,5 +241,178 @@
+        // depth = 1 and the presence of a mapperFunction guaranteed
+        // both these conditions are always met when this is called from flatMap
+        // Additionally this is slightly refactored rather than taking a thisArg
+        // the calling funciton binds the thisArg if it's required
```
nit: spelling in these comment lines

dilijev commented on lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js:
```diff
@@ -238,5 +241,178 @@
+                    while (innerIndex < innerLength) {
+                        if (innerIndex in element) {
+                            //  1. If targetIndex >= 2^53-1, throw a TypeError exception.
+                            if (targetIndex >= 9007199254740991 /* 2^53 -1 */) {
```
nit: remove space before `-1` for consistency

dilijev commented on lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js:
```diff
@@ -238,5 +241,178 @@
+                } else {
+                    //  vi. Else,
+                    //      1. If targetIndex >= 2^53-1, throw a TypeError exception.
+                    if (targetIndex >= 9007199254740991 /* 2^53 -1 */) {
```
nit: remove space before `-1` for consistency

fatcerberus commented on lib/Runtime/Library/JsBuiltIn/JsBuiltIn.js:
```diff
@@ -238,5 +241,164 @@
+            sourceLen =  __chakraLibrary.GetLength(o);
+        }
+        //3. If IsCallable(mapperFunction) is false throw a TypeError exception
+        if (!__chakraLibrary.isCallable(mapperFunction)) {
```
I believe that, per spec, `typeof x === 'function'` is exactly equivalent to `IsCallable(x)`.

ljharb replied:
For an engine, this is correct. In libraries that need to account for older engines where this does not hold, they'd need a more intensive check - but chakra shouldn't need to worry about that.

rhuanjl commented:
@dotnet-bot test OSX osx_osx_debug_static please

rhuanjl commented:
@dotnet-bot test OSX static_osx_osx_debug please

rhuanjl commented:
Please let me know what's needed for this to be merged.

I assume it will require a GUID update as it changes the byte code, anything else?

sigatrev commented:
@rhuanjl I think this is pretty much ready. I should be able to get it merged in the next couple of days, sorry for the delay.

sigatrev commented:
@rhuanjl can you enable us to update the PR so I update bytecode headers and GUID?

rhuanjl commented:
@sigatrev Allow edits from maintainers is already ticked not sure if there's anything else I can do to give you additional access?

(let me know if you'd like me to regen the byte code)

rhuanjl commented:
Thanks for testing and merging this @sigatrev 

And thanks everyone for the reviews.

sigatrev commented:
@rhuanjl Thanks for your patience. We really appreciate the contributions.
